pkg/events/watcher.go

```diff
@@ -33,161 +27,83 @@ type EventHandler interface {
-	}
-
-	w.ctl = ctl
-	w.waitForCacheSync = mgr.GetCache().WaitForCacheSync
-
-	return w, nil
+func NewWatcher(name string, mgr manager.Manager, resource runtime.Object) *watcher {
+	return &watcher{name: name, mgr: mgr, resource: resource}
 }
 
-func (w *watcher) getGvk(resource runtime.Object) (schema.GroupVersionKind, error) {
-	return apiutil.GVKForObject(resource, w.scheme)
-}
-
-func (w *watcher) getHandlers(resource runtime.Object) ([]EventHandler, error) {
-	gvk, err := w.getGvk(resource)
-	if err != nil {
-		return nil, err
-	}
-	w.lock.RLock()
-	handlers, ok := w.handlers[gvk]
-	w.lock.RUnlock()
-	if !ok {
-		return nil, errors.Errorf("no handler registered for gvk %v", gvk)
+func (w *watcher) Watch(ctx context.Context, eventHandler EventHandler, predicates ...predicate.Predicate) error {
+	reconciler := &eventWatcher{
+		events:       NewCache(),
+		eventHandler: eventHandler,
 	}
 
-	return handlers, nil
-}
-
-func (w *watcher) Watch(ctx context.Context, resource runtime.Object, eventHandler EventHandler, predicates ...predicate.Predicate) error {
-	// create a source for the resource type
-	src := &source.Kind{Type: resource}
-
-	var err error
-	w.startWatch.Do(func() {
-		// send watch events to the Cache
-		err = w.ctl.Watch(src, w.events, predicates...)
+	ctl, err := controller.New(w.name, w.mgr, controller.Options{
+		Reconciler: reconciler,
 	})
 	if err != nil {
 		return err
 	}
 
-	gvk, err := w.getGvk(resource)
-	if err != nil {
+	// create a source for the resource type
+	src := &source.Kind{Type: w.resource}
+
+	// send watch events to the Cache
+	if err := ctl.Watch(src, reconciler.events, predicates...); err != nil {
 		return err
 	}
 
-	// add the handler to our map
-	w.lock.Lock()
-	// use gvk as the key
-	handlers := w.handlers[gvk]
-	handlers = append(handlers, eventHandler)
-	w.handlers[gvk] = handlers
-	w.lock.Unlock()
-
-	if synced := w.waitForCacheSync(ctx.Done()); !synced {
+	if synced := w.mgr.GetCache().WaitForCacheSync(ctx.Done()); !synced {
 		return errors.Errorf("waiting for cache sync failed")
 	}
 
 	return nil
 }
 
-func (w *watcher) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+type eventWatcher struct {
+	events       Cache
+	eventHandler EventHandler
+}
+
+func (w *eventWatcher) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	// event key is stored in the request name
 	key := request.Name
-	log.Log.V(4).Info("watcher reconciling key", "key", key)
+	log.Log.V(4).Info("event eventWatcher reconciling event", "key", key)
 
 	event := w.events.Get(key)
 
 	if event == nil {
 		return reconcile.Result{}, errors.Errorf("internal error: received invalid event key %v", key)
 	}
 
-	switch event.EventType {
-	case EventTypeCreate:
-		obj := event.CreateEvent.Object
-		handlers, err := w.getHandlers(obj)
-		if err != nil {
+	switch evt := event.(type) {
+	case createEvent:
+		if err := w.eventHandler.Create(evt.Object); err != nil {
 			return reconcile.Result{}, err
```

Comment:
To help me understand, what's the reason why we're never requeueing events?

Reply:
if error is returned, it will requeue the event; however I'm planning on deprecating the event handler (which includes the watcher). the new reconciler will expose the reconcile.Result if you'd like to requeue without returning error